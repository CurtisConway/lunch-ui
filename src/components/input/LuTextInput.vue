<template>
  <div :class="defaultClassList">
    <LuText tag="label" :for="computedId" size="subtitle" :class="labelClassList">
      {{ label }}<span v-if="required">*</span>
    </LuText>
    <input
      ref="inputElement"
      :id="computedId"
      :type="inputType"
      v-model="valueInterface"
      @change="onChange"
      @focus="onFocus"
      @blur="onBlur"
    >
    <transition name="grow-fade">
      <LuErrorList :errors="errorList" v-if="!isValid && focus" />
    </transition>
  </div>
</template>

<script>
import './LuInput.scss';
import LuText from '../text/LuText.vue';
import LuErrorList from './partials/_LuErrorList.vue';
import colors from '../../assets/js/mixins/colors';
import borders from '../../assets/js/mixins/borders';
import spacing from '../../assets/js/mixins/spacing';
import events from '../../assets/js/mixins/events';
import inputs from './_inputs.mixin';
import validation from './_validation.mixin';

export default {
  name: 'LuTextInput',
  mixins: [colors, borders, spacing, events, inputs, validation],
  components: {
    LuText,
    LuErrorList,
  },
  props: {
    type: {
      type: String,
      default: () => 'text',
    },
  },
  computed: {
    inputType() {
      if (['text', 'password', 'email', 'tel', 'number', 'search'].includes(this.type)) {
        return this.type;
      }
      return 'text';
    },
  },
};
</script>
