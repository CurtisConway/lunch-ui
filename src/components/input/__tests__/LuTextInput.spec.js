import Vue from 'vue';
import Router from 'vue-router';

import LuTextInput from '../LuTextInput.vue';

import {
  createLocalVue,
  mount,
} from '@vue/test-utils';
import { compileToFunctions } from 'vue-template-compiler'

describe('LuTextInput.vue', () => {
  let mountFunction;
  let router = Router;
  let localVue = typeof Vue;

  beforeEach(() => {
    router = new Router();
    localVue = createLocalVue();
    localVue.use(Router);

    mountFunction = (options = {}) => {
      return mount(LuTextInput, {
        localVue,
        router,
        ...options,
      });
    }
  });

  it('should render component and match snapshot', () => {
    expect(mountFunction().html()).toMatchSnapshot();
  });

  it('should render with type password and match snapshot', () => {
    expect(mountFunction({
      propsData: {
        type: 'password',
      },
    }).html()).toMatchSnapshot();
  });


  it('should render with a label prop and match snapshot', () => {
    expect(mountFunction({
      propsData: {
        label: 'test',
      },
    }).html()).toMatchSnapshot();
  });

  it('should render with an inputStyle prop and match snapshot', () => {
    expect(mountFunction({
      propsData: {
        inputStyle: 'underline',
      },
    }).html()).toMatchSnapshot();

    expect(mountFunction({
      propsData: {
        inputStyle: 'solid',
      },
    }).html()).toMatchSnapshot();

    expect(mountFunction({
      propsData: {
        inputStyle: 'solo',
      },
    }).html()).toMatchSnapshot();
  });

  it('should render with an errors list and match snapshot', async () => {
    const wrapper = mountFunction({
      propsData: {
        errors: ['error one', 'error two'],
      },
    });
    wrapper.vm.$refs.inputElement.focus();
    await wrapper.vm.$nextTick();

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render with a success prop and match snapshot', () => {
    expect(mountFunction({
      propsData: {
        success: true,
      },
    }).html()).toMatchSnapshot();
  });

  it('should render with a persistentLabel prop and match snapshot', () => {
    expect(mountFunction({
      propsData: {
        persistentLabel: true,
      },
    }).html()).toMatchSnapshot();
  });

  it('should pass the value prop to the input element', async () => {
    const wrapper = mountFunction({
      propsData: {
        value: 'test',
      },
    });

    expect(wrapper.vm.$refs.inputElement.value).toBe('test');
  });

  it('should emit a focus event', async () => {
    const wrapper = mountFunction({
      propsData: {
        value: 'test',
      },
    });

    const event = jest.fn();
    wrapper.vm.$on('focus', event);
    wrapper.vm.$refs.inputElement.focus();
    wrapper.vm.$refs.inputElement.blur();
    wrapper.vm.$refs.inputElement.focus();
    wrapper.vm.$refs.inputElement.blur();

    expect(event.mock.calls).toHaveLength(2);
  });

  it('should emit a blur event', async () => {
    const wrapper = mountFunction({
      propsData: {
        value: 'test',
      },
    });

    const event = jest.fn();
    wrapper.vm.$on('blur', event);
    wrapper.vm.$refs.inputElement.focus();
    wrapper.vm.$refs.inputElement.blur();
    wrapper.vm.$refs.inputElement.focus();
    wrapper.vm.$refs.inputElement.blur();

    expect(event.mock.calls).toHaveLength(2);
  });

  it('should emit an input event', async () => {
    const wrapper = mountFunction({
      propsData: {
        value: 'test',
      },
    });

    const event = jest.fn();
    wrapper.vm.$on('input', event);
    wrapper.vm.$refs.inputElement.focus();
    wrapper.vm.$refs.inputElement.value = 'testing';
    wrapper.vm.$refs.inputElement.blur();
    wrapper.vm.$refs.inputElement.focus();
    wrapper.vm.$refs.inputElement.value = 'test';
    wrapper.vm.$refs.inputElement.blur();

    expect(event.mock.calls).toHaveLength(2);
  });

  it('should be valid without any validation rules', async () => {
    const wrapper = mountFunction({
      propsData: {
        rules: [],
      },
    });

    expect(wrapper.vm.isValid).toBe(true);
  });

  it('should not be valid with validation rules that fail', async () => {
    const wrapper = mountFunction({
      propsData: {
        rules: [v => !!v || 'Required'],
      },
    });
    wrapper.vm.validateInput();
    expect(wrapper.vm.isValid).toBe(false);
  });

  it('should be valid with validation rules that pass', async () => {
    const wrapper = mountFunction({
      propsData: {
        value: 'test',
        rules: [v => !!v || 'Required'],
      },
    });

    expect(wrapper.vm.isValid).toBe(true);
  });

  it('should be not valid after input if validation fails', async () => {
    const wrapper = mountFunction({
      propsData: {
        value: 'test',
        rules: [v => !!v || 'Required'],
      },
    });
    const textInput = wrapper.find('input[type="text"]');
    textInput.setValue('');
    wrapper.vm.validateInput();

    expect(wrapper.vm.isValid).toBe(false);
  });
});
