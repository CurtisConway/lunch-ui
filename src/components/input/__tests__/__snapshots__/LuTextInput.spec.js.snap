// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LuTextInput.vue should render component and match snapshot 1`] = `
<div class="lu-form-group text-black border-black lu-input-outline"><label class="lu-text text-subtitle text-left" for="input-1">
    <!----></label> <input id="input-1" type="text">
  <!---->
</div>
`;

exports[`LuTextInput.vue should render with a label prop and match snapshot 1`] = `
<div class="lu-form-group text-black border-black lu-input-outline"><label class="lu-text text-subtitle text-left" for="input-9">
    test
    <!----></label> <input id="input-9" type="text">
  <!---->
</div>
`;

exports[`LuTextInput.vue should render with a persistentLabel prop and match snapshot 1`] = `
<div class="lu-form-group text-black border-black lu-input-outline"><label class="lu-text text-subtitle text-left persistent" for="input-34">
    <!----></label> <input id="input-34" type="text">
  <!---->
</div>
`;

exports[`LuTextInput.vue should render with a success prop and match snapshot 1`] = `
<div class="lu-form-group text-green border-green lu-input-outline"><label class="lu-text text-subtitle text-left" for="input-30">
    <!----></label> <input id="input-30" type="text">
  <!---->
</div>
`;

exports[`LuTextInput.vue should render with an errors list and match snapshot 1`] = `
<div class="lu-form-group text-red border-red lu-input-outline"><label class="lu-text text-subtitle text-left focus" for="input-25">
    <!----></label> <input id="input-25" type="text">
  <ul class="lu-error-list bg-red text-white pv-1 pr-2 pl-5 rad-1 grow-fade-enter grow-fade-enter-active">
    <li class="mv-1">
      error one
    </li>
    <li class="mv-1">
      error two
    </li>
  </ul>
</div>
`;

exports[`LuTextInput.vue should render with an inputStyle prop and match snapshot 1`] = `
<div class="lu-form-group text-black border-black lu-input-underline"><label class="lu-text text-subtitle text-left" for="input-13">
    <!----></label> <input id="input-13" type="text">
  <!---->
</div>
`;

exports[`LuTextInput.vue should render with an inputStyle prop and match snapshot 2`] = `
<div class="lu-form-group text-black border-black lu-input-solid"><label class="lu-text text-subtitle text-left" for="input-17">
    <!----></label> <input id="input-17" type="text">
  <!---->
</div>
`;

exports[`LuTextInput.vue should render with an inputStyle prop and match snapshot 3`] = `
<div class="lu-form-group text-black border-black lu-input-solo elevation-3"><label class="lu-text text-subtitle text-left" for="input-21">
    <!----></label> <input id="input-21" type="text">
  <!---->
</div>
`;

exports[`LuTextInput.vue should render with type password and match snapshot 1`] = `
<div class="lu-form-group text-black border-black lu-input-outline"><label class="lu-text text-subtitle text-left" for="input-5">
    <!----></label> <input id="input-5" type="password">
  <!---->
</div>
`;
